Highlight

willkg opened the pull request:
I'm pretty sure this implements the great plan.  There are a couple of deviations....
1. content_for_fields always returns a tuple of content as specified and in the order of highlight_fields
2. highlight_fields is no longer optional--you need to explicitly state which fields you want highlighted

f?

willkg committed:
Adds oedipus.egg-info to .gitignore

willkg committed:
5. Adds highlight/excerpt

erikrose commented on oedipus/__init__.py:
```diff
@@ -132,6 +136,34 @@ class S(object):
+        Note: This won't let you clear it--we'd need to write a
+        ``clear_highlight()``.
+
+        :param highlight_fields: The list of fields to highlight.  This
```
Just commenting as I go here, so this may make more sense to me later, but…why not use `*highlight_fields` rather than making the caller provide brackets?

willkg replied:
Doing *args is sometimes nice, but it's generally bad to do when mixed with additional arguments.

erikrose replied:
But the rest are kwargs, so I wouldn't anticipate any confusion. Do you forsee more non-kw args? Django's order_by() takes *args, so there's a precedent.

willkg replied:
In order to do what you want, you have to switch the signature to:

```
def highlight(self, *args, **kwargs):
```

Is that what you want to change this to?  You lose some of the implicit documentation this way.  It doesn't seem nicer to me.

erikrose replied:
I'm with you on the self-documentation. It's too bad that kwargs and default arg values are mixed together in Python, but we can't really do anything about that now.

Though I'm not excited about what *args would do to the implementation, I'd like for highlight() to have the same calling conventions as `order_by()`, `values()`, and `values_dict()`. A bunch of calls to `kwargs.pop()` at the top of the method (and your excellent Sphinx docs) can keep the un-self-documenting-ness to a minimum.

erikrose commented on oedipus/__init__.py:
```diff
@@ -132,6 +136,34 @@ class S(object):
+        :param before_match: The html tag for before the match.
+        :param after_match: The html tag for after the match.
+        :param limit: The maximum snippet size in symbols.  Defaults
+            to ``settings.SEARCH_SUMMARY_LENGTH``.
```
And maybe later we can let you put some defaults on the SphinxMeta if we need it.

willkg replied:
This default should totally be in SphinxMeta.  That allows you to set limits on an thing-by-thing basis.  I'll move it.

erikrose replied:
I wouldn't want to make every definer of a SphinxMeta define a SEARCH_SUMMARY_LENGTH if they don't care, so it wouldn't be a bad thing to have a global to fall back to.

willkg replied:
Sphinx defaults to 256.  settings.SEARCH_SUMMARY_LENGTH in Kitsune is set to 275.

I think I'll fix it so if it's passed in as an arg, that's used.  If it's not and there's a SphinxMeta, that's used.  If it's not there either, then it doesn't specify it at all and the Sphinx default limit is used.

erikrose replied:
Ideal!

erikrose commented on oedipus/__init__.py:
```diff
@@ -132,6 +136,34 @@ class S(object):
+        :returns: The new ``S``.
+
+        """
+        if limit == None:
```
Why not `is None`?

erikrose replied:
For that matter, why can't we just put `limit=settings.SEARCH_SUMMARY_LENGTH` in the arg list?

willkg replied:
The "== None" is probably from an a search-and-replace.

It should get changed and setting it to settings.SEARCH_SUMMARY_LENGTH (or the future SphinxMeta.excerpt_limit) is a good idea.

erikrose commented on oedipus/__init__.py:
```diff
@@ -219,6 +251,28 @@ class S(object):
     __len__ = count
 
+    def excerpt(self, result):
+        """Takes a result and returns the excerpt as a string."""
```
nit: that variant present tense which says "take" and "return" for docstrings, as per PEP 257, pretty please :-)

willkg replied:
I think this is nit-picky in a way that doesn't affect tools at all.  Do we really want to be this nit-picky?  We're not this nit-picky about other things (e.g. git commit first line length, docstring first line length (also in PEP 257)) which have affects the tools we use.

I'll change it because you asked (and will do that in the future, too), but it's not something I think is helpful to spend time caring about.

erikrose replied:
I confess to having something of a heightened sensitivity where matters of grammatical consistency are concerned. ;-)

erikrose commented on oedipus/__init__.py:
```diff
@@ -219,6 +251,28 @@ class S(object):
+        options = {}
+        for opt in ('before_match', 'after_match', 'limit'):
+            if opt in self._highlight_options:
+                options[opt] = self._highlight_options[opt]
```
In Python 2.7, we could probably have used set ops on dictionary view or something, resulting in something like `option.update(..., except_for=this_key)`. Maybe someday. :-)

willkg replied:
Yeah, I explicitly didn't do an exclusion-style update.  Doing it this way makes it clear exactly what's in the resulting dict at the point it's created rather than at a point farther away in the code.  It'll be easier to add options to the highlight_options dict without accidentally affecting the options we're sending to Sphinx.

However, currently, this is easily done in a one-liner:

options = dict((opt, self._highlight_options[opt]) for opt in ('before_match', 'after_match', 'limit'))

I'll change it to that.  We'll have to revert it if we ever end up with highlight_options that don't have exact matches with the options we send to Sphinx in BuildExcerpts.

erikrose replied:
Either way. And I share your philosophy about explicitness and future-proofing.

erikrose commented on oedipus/__init__.py:
```diff
@@ -219,6 +251,28 @@ class S(object):
+        if self._results_cache is None:
+            # This guarantees that we have results before we try
+            # getting excerpts for results.
+            self._raw()
```
I'll have to look and see if this is really necessary. Have you tried without? I would naively hope that any lazy stuff would resolve as necessary, but I could conceive of situations where it wouldn't.

willkg replied:
It's kind of non-trivial, but you can kick up a KeyError if you call excerpt on an S with results from another S.  One shouldn't be doing that sort of thing, but getting a KeyError would be puzzling.  I figured it's an easy/fast check to see if we've already got results or not.

Having said that, this is all kind of weird and if you're in this situation, it's better to throw an error rather than let it slide through and get an excerpt with settings from a different S.  So maybe instead of calling self._raw(), it's better to throw a DudeThisSHasNoResultsYet error.  What do you think?

erikrose replied:
If it's not too much work, let's throw a DudeThisSHasNoResultsYet error. Hmm, now you've got my second-guessing our decision to not put excerpt() on the search result somehow....

erikrose commented on oedipus/results.py:
```diff
@@ -70,3 +108,22 @@ class ObjectResults(SearchResults):
+        :returns: Tuple with content in the field indexes specified by
+            ``highlight_fields``.
+
+        :raises AttributeError: If there is a field in
```
Soooo, all 3 implementations of `content_for_fields` share basically the same docstring, except for this line. What if we parked this docstring in an abstract method in the superclass and just wrapped these implementations to catch the 3 different errors and raise a common `NoSuchFieldException` or something? Folks from the outside really shouldn't be calling this method anyway.

willkg replied:
It feels icky to be doing subclassing kinds of things to reduce redundant documentation.  If you dislike the docs, I can remove them.

erikrose commented on oedipus/results.py:
```diff
@@ -70,3 +108,22 @@ class ObjectResults(SearchResults):
+            ``highlight_fields`` that doesn't exist in ``fields``.
+
+        """
+        content = [getattr(result, field) for field in highlight_fields]
```
Couldn't we just say `tuple(getattr(result, field) for field in highlight_fields)`?

willkg replied:
Yeah.  I was leaving it the "long way" to make it clearer what's going on so it's easier to give feedback on.  I'll go through and tighten things up.

erikrose replied:
I actually generally find the shorter ways more readable, for future reference, so no need to stretch things out on my account!

willkg commented:
I marked this as an f? and was kind of hoping to get some feedback covering the deviations and whether this accomplishes what we set out to accomplish on Thursday.  While the review criticisms are good, I had tried to make the code trivially readable to make it easier to verify that it's doing the right thing.

Am I right that f? is for feedback regarding approach and things like that and r? is for review with the intention of merging it in next?

erikrose commented:
f+ :-) I l think the approach is sound.

willkg commented:
I'm going through and addressing the review issues.

willkg committed:
Fixes issues

* moves summary_length to SphinxMeta and doesn't pass a limit at all if
  it's not specified (which will default to Sphinx's 256 limit)
* adds test for that
* adds test for making sure _raw gets called
* cleans up and tightens some code
* fixes tense in docstring
* added additional comment about excerpt() getting called before results
  have been evaluated

willkg commented:
Erik: Can I get an review? on the changes?

erikrose commented:
Did we ever do anything about the case in which the fields the caller asks for don't include the highlit ones? If not, we should at least file a bug about that.

erikrose commented on oedipus/tests/test_excerpt.py:
```diff
@@ -0,0 +1,140 @@
+    s._raw()
+
+    # These three are set in the highlight() call.
+    eq_(s._highlight_options['before_match'], '<i>')
```
LIkewise, we can do one eq_() for all 3 here.

erikrose commented on oedipus/tests/test_excerpt.py:
```diff
@@ -0,0 +1,140 @@
+    s._raw()
+
+    # There should be no limit defined.
+    eq_('limit' in s._highlight_options, False)
```
`assert 'limit' not in s._highlight_options`. eq_() is really just for when it's handy to see the value when it fails.

erikrose commented on oedipus/tests/test_excerpt.py:
```diff
@@ -0,0 +1,140 @@
+@fudge.patch('sphinxapi.SphinxClient')
+@fudge.patch('oedipus.tests.Biscuit')
+def test_highlight_chainable_no_sphinx_meta_limit(sphinx_client, biscuit):
+    """Test highlight chainable sets options."""
```
This has the same docstring as the previous one.

erikrose commented on oedipus/tests/test_excerpt.py:
```diff
@@ -0,0 +1,140 @@
+    (sphinx_client.expects_call().returns_fake()
+                  .is_a_stub()
+                  .expects('BuildExcerpt')
+                  .with_args(('sesame', 'has sesame foo'), 'biscuit',
```
Oh god the inscrutable wrapping! (You don't have to change it, but I have to complain about it. ;-))

erikrose commented on oedipus/tests/test_excerpt.py:
```diff
@@ -0,0 +1,140 @@
+
+@fudge.patch('sphinxapi.SphinxClient')
+def test_excerpt_forces_results(sphinx_client):
+    """Test excerpt forces _raw() to get called."""
```
I'm very glad you tested this. :-D

erikrose commented on oedipus/tests/test_excerpt.py:
```diff
@@ -0,0 +1,140 @@
+@fudge.patch('sphinxapi.SphinxClient')
+def test_excerpt_forces_results(sphinx_client):
+    """Test excerpt forces _raw() to get called."""
+    Biscuit(id=123, name='sesame', content='has sesame foo')
```
Whenever you're instantiating biscuits, you need to do something to clear the mock model_cache afterward. I have a SphinxMockingTestCase up in `__init__` that comes in handy.

erikrose commented on oedipus/tests/test_excerpt.py:
```diff
@@ -0,0 +1,140 @@
+
+@fudge.patch('sphinxapi.SphinxClient')
+def test_excerpt_limit_fields(sphinx_client):
+    """Test excerpt with some fields highlighted."""
```
I had trouble figuring out how this was different from test_excerpt(), above. Saying "_only_ some fields" would clear it right up for Future Erik.

willkg committed:
More fixes

* reworks highlight() so it's like the other chainables
* reworks defaults so that all defaults can be specified on SphinxMeta
  with ``excerpt_`` attributes
* creates ExcerptError
* fixes excerpt so it makes sure highlight_fields is a subset of fields
  and adds test for it
* cleans up some silly looking test code
* adds test to make sure highlight options override previous options

willkg commented:
Sorry about this Erik.  This last commit was a bunch of non-trivial changes.  If you want, I can squash and push and that'll make it easier to see all the changes holistically.

Pretty sure I addressed everything.

willkg commented:
And by "holistically", I really mean "as a whole and not piecemeal"---"holistically" is totally the wrong word to use there.

erikrose commented:
"Holistic" seems quite appropriate! "characterized by comprehension of the parts of something as intimately interconnected and explicable only by reference to the whole"

erikrose commented:
Looking good! Just a few last comments.

willkg committed:
More more fixes

* fixes some " usage
* fixes exception messages: one was obtuse and one was formatted wrogn
* puts Biscuit building test cases in a TestCase class that has a
  tearDown that cleans things up properly
* tidies some test code
* unfortunately, I also moved a bunch of test code around so it's not
  changed, but looks changed

willkg commented:
I think I covered everything left with this one plus a couple of additional things I caught.

Last review?.

willkg committed:
Changes highlight_fields to a *args argument

willkg commented:
That covers the *fields thing in highlight.  Review?

erikrose commented:
r+! Nice work; can't wait to have it merged in.

willkg commented:
f0f4fd71e0401ebbe543a4af2dd583c3d85ea115
